app/src/middleware/carboneCopy.js

```diff
@@ -0,0 +1,251 @@
+const config = require('config');
+const log = require('npmlog');
+const moment = require('moment');
+const morgan = require('morgan');
+const path = require('path');
+
+const keycloak = require('../components/keycloak');
+
+const clientId = config.get('keycloak.clientId');
+
+const operations = Object.freeze([
+  {method: 'POST', regex: /\/template$/, name: 'UPLOAD_TEMPLATE', isGenerator: false},
+  {method: 'POST', regex: /\/template\/\w+\/render/g, name: 'GENERATE_FROM_TEMPLATE', isGenerator: true},
+  {method: 'POST', regex: /\/template\/render$/, name: 'GENERATE_FROM_UPLOAD', isGenerator: true},
+  {method: 'GET', regex: /\/render/g, name: 'GET_OUTPUT', isGenerator: false},
+  {method: 'GET', regex: /\/template\/(?!render)/g, name: 'GET_TEMPLATE', isGenerator: false}
+]);
+
+const getOperation = (req) => {
+  const result = {isTrackable: false, isGenerator: false, name: null};
+  if (!req) return result;
+
+  operations.forEach(o => {
+    if (!result.operation) {
+      if (req.method === o.method && req.url.match(o.regex)) {
+        result.name = o.name;
+        result.isTrackable = true;
+        result.isGenerator = o.isGenerator;
+      }
+    }
+  });
+
+  return result;
+};
+
+const apiFormat = ':op :azp :ts :status :response-time';
+const generatorApiFormat = `${apiFormat} :outputFileType :contextKeyCount :contentFileType :contentEncodingType :contentSize :res[content-length]`;
+
+const apiTracker = async (req, res, next) => {
+  const operation = req._carboneOp;
+
+  if (operation && operation.isTrackable) {
+    req._ts = moment.utc().valueOf();
+    req._op = operation.name;
+  }
+  next();
+};
+
+const apiTrackerParse = (msg) => {
+  if (msg && msg.trim().length) {
+    const parts = msg.trim().split(' ');
+
+    const ms = Number.parseFloat(parts[4]);
+    const s = ms / 1000;
+    const roundedSeconds = Math.round(s * 10) / 10;
+
+    let message = {
+      type: 'api-tracker',
+      timestamp: Number.parseInt(parts[2]),
+      op: parts[0],
+      azp: parts[1],
+      response: {
+        status: Number.parseInt(parts[3]),
+        timeMs: s,
+        seconds: roundedSeconds
+      }
+    };
+
+    let extended = {};
+    if (parts.length === 11) {
+      const outputFileType = parts[5];
+      const contextKeyCount = Number.parseInt(parts[6]);
+      const contentFileType = parts[7];
+      const contentEncodingType = parts[8];
+      const contentSize = Number.parseInt(parts[9]);
+      const contentLength = '-' !== parts[10] ? Number.parseInt(parts[10]) : undefined;
+      extended = {
+        context: {
+          keyCount: contextKeyCount
+        },
+        content: {
+          fileType: contentFileType,
+          encodingType: contentEncodingType,
+          size: contentSize
+        },
+        output: {
+          fileType: outputFileType,
+          size: contentLength
+        }
+      };
```

Comment:
It is possible to have extended declared as a constant if you did object assignment instead. For example:
```
const extended = {};
...
extended.context = { keyCount: Number.parseInt(parts[6]) };
extended.content = { ... };
extended.output = { ... };
```

Another advantage of this is that you can inline more of the declarations directly in the object declaration. You can avoid having to make a bunch of const declarations from lines 71 to 76 which iirc is slightly more memory and cpu efficient.

Reply:
very cool, done.